File: freightmas/freightmas/report/statement_of_accounts/statement_of_accounts.py
# Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import formatdate

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "posting_date",
            "label": _("Date"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "voucher_type",
            "label": _("Voucher Type"),
            "fieldtype": "Data",
            "width": 130
        },
        {
            "fieldname": "voucher_no",
            "label": _("Voucher No"),
            "fieldtype": "Dynamic Link",
            "options": "voucher_type",
            "width": 195,
            "align": "left"
        },
        {
            "fieldname": "debit",
            "label": _("Debit"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "credit",
            "label": _("Credit"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "balance",
            "label": _("Balance"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "remarks",
            "label": _("Remarks"),
            "fieldtype": "Text",
            "width": 340
        }
    ]

def get_data(filters):
    data = []
    
    # Get opening balance
    opening_balance = get_opening_balance(filters)
    balance = opening_balance
    
    # Add opening balance row
    data.append({
        "posting_date": formatdate(filters.get("from_date"), "dd-MMM-yy"),
        "voucher_type": "Opening Balance",
        "voucher_no": "",
        "debit": 0,
        "credit": 0,
        "balance": opening_balance,
        "remarks": "Opening Balance"
    })
    
    conditions = get_conditions(filters)
    
    # Get GL entries for the party
    gl_entries = frappe.db.sql("""
        SELECT 
            posting_date,
            voucher_type,
            voucher_no,
            account,
            party,
            debit,
            credit,
            remarks
        FROM `tabGL Entry`
        WHERE party_type = %(party_type)s
        AND {conditions}
        ORDER BY posting_date, creation
    """.format(conditions=conditions), filters, as_dict=1)

    # Get draft invoices if requested
    if filters.get("include_draft_invoices"):
        draft_invoices = get_draft_invoices(filters)
        all_entries = gl_entries + draft_invoices
        all_entries = sorted(all_entries, key=lambda x: x.posting_date)
    else:
        all_entries = gl_entries

    for entry in all_entries:
        balance += (entry.debit - entry.credit)
        row = {
            "posting_date": formatdate(entry.posting_date, "dd-MMM-yy"),
            "voucher_type": entry.voucher_type,
            "voucher_no": entry.voucher_no,
            "debit": entry.debit,
            "credit": entry.credit,
            "balance": balance,
            "remarks": entry.remarks
        }
        data.append(row)
    
    # Add closing balance row
    data.append({
        "posting_date": formatdate(filters.get("to_date"), "dd-MMM-yy"),
        "voucher_type": "Closing Balance",
        "voucher_no": "",
        "debit": 0,
        "credit": 0,
        "balance": balance,
        "remarks": "Closing Balance"
    })
    
    return data

def get_opening_balance(filters):
    balance = frappe.db.sql("""
        SELECT SUM(debit) - SUM(credit) as balance
        FROM `tabGL Entry`
        WHERE party_type = %(party_type)s
        AND posting_date < %(from_date)s
        AND company = %(company)s
        AND party = %(party)s
    """, filters, as_dict=1)[0].balance
    
    return balance or 0

def get_draft_invoices(filters):
    if filters.get("party_type") == "Customer":
        return get_draft_sales_invoices(filters)
    else:
        return get_draft_purchase_invoices(filters)

def get_draft_sales_invoices(filters):
    return frappe.db.sql("""
        SELECT 
            posting_date,
            'Sales Invoice' as voucher_type,
            name as voucher_no,
            '' as account,
            customer as party,
            grand_total as debit,
            0 as credit,
            CONCAT('Draft Invoice - ', name) as remarks
        FROM `tabSales Invoice`
        WHERE docstatus = 0
        AND customer = %(party)s
        AND company = %(company)s
        AND posting_date >= %(from_date)s
        AND posting_date <= %(to_date)s
    """, filters, as_dict=1)

def get_draft_purchase_invoices(filters):
    return frappe.db.sql("""
        SELECT 
            posting_date,
            'Purchase Invoice' as voucher_type,
            name as voucher_no,
            '' as account,
            supplier as party,
            0 as debit,
            grand_total as credit,
            CONCAT('Draft Invoice - ', name) as remarks
        FROM `tabPurchase Invoice`
        WHERE docstatus = 0
        AND supplier = %(party)s
        AND company = %(company)s
        AND posting_date >= %(from_date)s
        AND posting_date <= %(to_date)s
    """, filters, as_dict=1)

def get_conditions(filters):
    conditions = []
    
    if filters.get("company"):
        conditions.append("company = %(company)s")
    if filters.get("from_date"):
        conditions.append("posting_date >= %(from_date)s")
    if filters.get("to_date"):
        conditions.append("posting_date <= %(to_date)s")
    if filters.get("party"):
        conditions.append("party = %(party)s")
        
    return " AND ".join(conditions)

File: freightmas/freightmas/report/statement_of_accounts/statement_of_accounts.js
// Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
// For license information, please see license.txt

frappe.query_reports["Statement of Accounts"] = {
    onload: function(report) {
        report.page.add_inner_button('Export to Excel', function() {
            const filters = report.get_filter_values(true);
            const query = encodeURIComponent(JSON.stringify(filters));
            const url = `/api/method/freightmas.api.export_statement_of_accounts_to_excel?filters=${query}`;
            window.open(url);
        }, 'Export');

        report.page.add_inner_button('Export to PDF', function() {
            const filters = report.get_filter_values(true);
            const query = encodeURIComponent(JSON.stringify(filters));
            const url = `/api/method/freightmas.api.export_statement_of_accounts_to_pdf?filters=${query}`;
            window.open(url);
        }, 'Export');

        report.page.add_inner_button('Clear Filters', function() {
            report.filters.forEach(filter => {
                let default_value = filter.df.default || "";
                report.set_filter_value(filter.df.fieldname, default_value);
            });
            report.refresh();
        });
    },

    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "party_type",
            "label": __("Party Type"),
            "fieldtype": "Select",
            "options": ["Customer", "Supplier"],
            "default": "Customer",
            "reqd": 1,
            "on_change": function() {
                let party_type = frappe.query_report.get_filter_value('party_type');
                frappe.query_report.toggle_filter_display('party', true);
                frappe.query_report.set_filter_value('party', '');
                
                let party_field = frappe.query_report.get_filter('party');
                party_field.df.options = party_type;
                party_field.df.label = __(party_type);
                party_field.refresh();
            }
        },
        {
            "fieldname": "party",
            "label": __("Party"), // This will change dynamically based on party_type
            "fieldtype": "Link",
            "options": "Customer", // This will change dynamically based on party_type
            "reqd": 1,
            "get_query": function() {
                let party_type = frappe.query_report.get_filter_value('party_type');
                if (!party_type) {
                    frappe.throw(__("Please select Party Type first"));
                }
                return {
                    filters: {
                        'disabled': 0
                    }
                };
            }
        },
        {
            "fieldname": "date_range",
            "label": __("Date Range"),
            "fieldtype": "Select",
            "options": [
                "",
                "Today",
                "Yesterday",
                "This Week",
                "Last Week",
                "This Month",
                "Last Month",
                "This Year",
                "Last Year",
                "Custom"
            ],
            "default": "This Month",
            "on_change": function() {
                let date_range = frappe.query_report.get_filter_value('date_range');
                let today = frappe.datetime.get_today();
                let from_date = '';
                let to_date = '';

                switch (date_range) {
                    case "Today":
                        from_date = to_date = today;
                        break;
                    case "Yesterday":
                        from_date = to_date = frappe.datetime.add_days(today, -1);
                        break;
                    case "This Week":
                        from_date = frappe.datetime.week_start();
                        to_date = frappe.datetime.week_end();
                        break;
                    case "Last Week":
                        from_date = frappe.datetime.add_days(frappe.datetime.week_start(), -7);
                        to_date = frappe.datetime.add_days(frappe.datetime.week_end(), -7);
                        break;
                    case "This Month":
                        from_date = frappe.datetime.month_start();
                        to_date = frappe.datetime.month_end();
                        break;
                    case "Last Month":
                        from_date = frappe.datetime.add_months(frappe.datetime.month_start(), -1);
                        to_date = frappe.datetime.add_days(frappe.datetime.month_start(), -1);
                        break;
                    case "This Year":
                        from_date = frappe.datetime.year_start();
                        to_date = frappe.datetime.year_end();
                        break;
                    case "Last Year":
                        let last_year = parseInt(frappe.datetime.get_today().split("-")[0]) - 1;
                        from_date = last_year + "-01-01";
                        to_date = last_year + "-12-31";
                        break;
                }

                if (date_range && date_range !== "Custom" && date_range !== "") {
                    frappe.query_report.set_filter_value('from_date', from_date);
                    frappe.query_report.set_filter_value('to_date', to_date);
                }
            }
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_end(),
            "reqd": 1
        },
        {
            "fieldname": "include_draft_invoices",
            "label": __("Include Draft Invoices"),
            "fieldtype": "Check",
            "default": 0
        }
    ]
};
